File: Week3 (Linked List, Doubly Linked List, Hash)/Doubly-Linked-List/lbh_doubly_linked_list_without_sentinel.c
#include <stdio.h>
#include <stdlib.h>

// Node structure
typedef struct Node {
    int data;
    struct Node* next;
    struct Node* prev;
} Node;

// Doubly linked list structure
typedef struct {
    Node* head;
    Node* tail;
    int size;
} DoublyLinkedList;

// Function to initialize the doubly linked list
void initList(DoublyLinkedList* list) {
    list->head = NULL;
    list->tail = NULL;
    list->size = 0;
}

// Function to insert a node at the beginning of the list
void insertFirst(DoublyLinkedList* list, int value) {
    Node* newNode = (Node*)malloc(sizeof(Node));
    newNode->data = value;
    newNode->next = list->head;
    newNode->prev = NULL;
    if (list->head != NULL) {
        list->head->prev = newNode;
    } else {
        list->tail = newNode; // If list was empty, new node is also the tail
    }
    list->head = newNode;
    list->size++;
}

// Function to insert a node at the end of the list
void insertLast(DoublyLinkedList* list, int value) {
    Node* newNode = (Node*)malloc(sizeof(Node));
    newNode->data = value;
    newNode->next = NULL;
    newNode->prev = list->tail;
    if (list->tail != NULL) {
        list->tail->next = newNode;
    } else {
        list->head = newNode; // If list was empty, new node is also the head
    }
    list->tail = newNode;
    list->size++;
}

// Function to insert a node before a node with a certain value
void insertValue(DoublyLinkedList* list, int target, int value) {
    Node* newNode = (Node*)malloc(sizeof(Node));
    newNode->data = value;
    Node* temp = list->head;
    while (temp != NULL && temp->data != target) {
        temp = temp->next;
    }
    if (temp == NULL) {
        printf("Target value not found\n");
        free(newNode);
        return;
    }
    newNode->next = temp;
    newNode->prev = temp->prev;
    if (temp->prev != NULL) {
        temp->prev->next = newNode;
    } else {
        list->head = newNode; // Inserting before the first node
    }
    temp->prev = newNode;
    list->size++;
}

// Function to delete a node by value
void deleteValue(DoublyLinkedList* list, int value) {
    Node* temp = list->head;
    while (temp != NULL && temp->data != value) {
        temp = temp->next;
    }
    if (temp == NULL) {
        return; // Value not found
    }
    if (temp->prev != NULL) {
        temp->prev->next = temp->next;
    } else {
        list->head = temp->next; // Node to be deleted is head
    }
    if (temp->next != NULL) {
        temp->next->prev = temp->prev;
    } else {
        list->tail = temp->prev; // Node to be deleted is tail
    }
    free(temp);
    list->size--;
}

// Function to delete the first node
void deleteFirst(DoublyLinkedList* list) {
    if (list->head == NULL) {
        return;
    }
    Node* toDelete = list->head;
    list->head = list->head->next;
    if (list->head != NULL) {
        list->head->prev = NULL;
    } else {
        list->tail = NULL; // List became empty
    }
    free(toDelete);
    list->size--;
}

// Function to delete the last node
void deleteLast(DoublyLinkedList* list) {
    if (list->tail == NULL) {
        return;
    }
    Node* toDelete = list->tail;
    list->tail = list->tail->prev;
    if (list->tail != NULL) {
        list->tail->next = NULL;
    } else {
        list->head = NULL; // List became empty
    }
    free(toDelete);
    list->size--;
}

// Function to get the first node's value
int getFirst(const DoublyLinkedList* list) {
    if (list->head == NULL) {
        printf("List is empty\n");
        return -1;
    }
    return list->head->data;
}

// Function to get the last node's value
int getLast(const DoublyLinkedList* list) {
    if (list->tail == NULL) {
        printf("List is empty\n");
        return -1;
    }
    return list->tail->data;
}

// Function to get the size of the list
int getSize(const DoublyLinkedList* list) {
    return list->size;
}

// Function to print the elements of the list
void printList(const DoublyLinkedList* list) {
    Node* temp = list->head;
    while (temp != NULL) {
        printf("%d -> ", temp->data);
        temp = temp->next;
    }
    printf("NULL\n");
}

// Function to free all nodes in the list
void freeList(DoublyLinkedList* list) {
    Node* temp = list->head;
    while (temp != NULL) {
        Node* next = temp->next;
        free(temp);
        temp = next;
    }
    list->head = NULL;
    list->tail = NULL;
    list->size = 0;
}

// Main function to demonstrate usage
int main() {
    DoublyLinkedList list;
    initList(&list);

    insertFirst(&list, 10);
    insertFirst(&list, 20);
    insertLast(&list, 30);
    insertLast(&list, 40);

    printf("List: ");
    printList(&list);
    printf("Size: %d\n", getSize(&list));

    printf("First element: %d\n", getFirst(&list));
    printf("Last element: %d\n", getLast(&list));

    insertValue(&list, 30, 25);
    printf("After inserting 25 before 30: ");
    printList(&list);

    deleteFirst(&list);
    printf("After deleting first element: ");
    printList(&list);

    deleteLast(&list);
    printf("After deleting last element: ");
    printList(&list);

    deleteValue(&list, 30);
    printf("After deleting value 30: ");
    printList(&list);

    printf("Final size: %d\n", getSize(&list));

    freeList(&list);
    return 0;
}

File: Week3 (Linked List, Doubly Linked List, Hash)/Doubly-Linked-List/lbh_doubly_linked_list_without_sentinel.py
### Python Implementation (Doubly Linked List without Sentinel Nodes)
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None

class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0

    # Function to insert a node at the beginning of the list
    def insert_first(self, value):
        new_node = Node(value)
        new_node.next = self.head
        new_node.prev = None
        if self.head is not None:
            self.head.prev = new_node
        else:
            self.tail = new_node  # If list was empty, new node is also the tail
        self.head = new_node
        self.size += 1

    # Function to insert a node at the end of the list
    def insert_last(self, value):
        new_node = Node(value)
        new_node.next = None
        new_node.prev = self.tail
        if self.tail is not None:
            self.tail.next = new_node
        else:
            self.head = new_node  # If list was empty, new node is also the head
        self.tail = new_node
        self.size += 1

    # Function to insert a node before a node with a certain value
    def insert_value(self, target, value):
        new_node = Node(value)
        temp = self.head
        while temp is not None and temp.data != target:
            temp = temp.next

        if temp is None:
            print("Target value not found")
            return

        new_node.next = temp
        new_node.prev = temp.prev
        if temp.prev is not None:
            temp.prev.next = new_node
        else:
            self.head = new_node  # Inserting before the first node
        temp.prev = new_node
        self.size += 1

    # Function to delete a node by value
    def delete_value(self, value):
        temp = self.head
        while temp is not None and temp.data != value:
            temp = temp.next

        if temp is None:
            return

        if temp.prev is not None:
            temp.prev.next = temp.next
        else:
            self.head = temp.next  # Node to be deleted is head

        if temp.next is not None:
            temp.next.prev = temp.prev
        else:
            self.tail = temp.prev  # Node to be deleted is tail

        self.size -= 1

    # Function to delete the first node
    def delete_first(self):
        if self.head is None:
            return

        self.head = self.head.next
        if self.head is not None:
            self.head.prev = None
        else:
            self.tail = None  # List became empty

        self.size -= 1

    # Function to delete the last node
    def delete_last(self):
        if self.tail is None:
            return

        self.tail = self.tail.prev
        if self.tail is not None:
            self.tail.next = None
        else:
            self.head = None  # List became empty

        self.size -= 1

    # Function to get the first node's value
    def get_first(self):
        if self.head is None:
            print("List is empty")
            return -1
        return self.head.data

    # Function to get the last node's value
    def get_last(self):
        if self.tail is None:
            print("List is empty")
            return -1
        return self.tail.data

    # Function to get the size of the list
    def get_size(self):
        return self.size

    # Function to print the elements of the list
    def print_list(self):
        temp = self.head
        while temp is not None:
            print(temp.data, end=" -> ")
            temp = temp.next
        print("NULL")

    # Function to free all nodes in the list
    def free_list(self):
        self.head = None
        self.tail = None
        self.size = 0

# Demonstration
list = DoublyLinkedList()
list.insert_first(10)
list.insert_first(20)
list.insert_last(30)
list.insert_last(40)

print("List: ", end="")
list.print_list()
print("Size:", list.get_size())

print("First element:", list.get_first())
print("Last element:", list.get_last())

list.insert_value(30, 25)
print("After inserting 25 before 30: ", end="")
list.print_list()

list.delete_first()
print("After deleting first element: ", end="")
list.print_list()

list.delete_last()
print("After deleting last element: ", end="")
list.print_list()

list.delete_value(30)
print("After deleting value 30: ", end="")
list.print_list()

print("Final size:", list.get_size())